asyncpg/connection.py

```diff
@@ -286,14 +284,17 @@ def transaction(self, *, isolation='read_committed', readonly=False,
             stmt_name = ''
 
         statement = await self._protocol.prepare(stmt_name, query, timeout)
-
         ready = statement._init_types()
         if ready is not True:
-            if self._types_stmt is None:
-                self._types_stmt = await self.prepare(self._intro_query)
-
-            types = await self._types_stmt.fetch(list(ready))
+            types, intro_stmt = await self.__execute(
+                self._intro_query, (list(ready),), 0, timeout)
             self._protocol.get_settings().register_data_types(types)
+            if not intro_stmt.name and not statement.name:
```

Comment:
Just check for `statement.name`? The current check won’t work correctly if `statement` is anonymous and `intro_stmt` is not. Also, tests?

Reply:
I don't think that's true.  From the docs:

> An unnamed prepared statement lasts only until the next Parse statement specifying the unnamed statement as destination is issued.

I added another test to show that.